postproengine/spod.py

```diff
@@ -0,0 +1,685 @@
+
+                    for ktheta_ind , ktheta_val in enumerate(ktheta):
+                        for tfreq_ind , tfreq_val in enumerate(tfreq):
+
+                            POD_Mat = np.zeros((NR*len(corr_inds),NB),dtype=complex)
+                            for corr_ind_iter , corr_ind in enumerate(corr_inds):
+                                POD_Mat[corr_ind_iter*NR:NR*(corr_ind_iter+1),0:NB] = np.copy(udata_rhat[:,ktheta_ind,:,tfreq_ind,corr_ind])
+                            POD_Mat = np.sqrt(scaling_factor_k) * np.dot(Wsqrt,POD_Mat)
+                            
+                            if compute_eigen_vectors:
+                                lsvd, ssvd, rsvd = np.linalg.svd(POD_Mat,full_matrices=False,compute_uv=True)
+                                eigmodes = np.zeros((NR,NB,len(corr_inds)),dtype=complex)
+
+                                temp  = np.dot(Wsqrtinv,lsvd)
+                                for corr_ind_iter , corr_ind in enumerate(corr_inds):
+                                    eigmodes[:,:,corr_ind_iter] = temp[corr_ind_iter*NR:NR*(corr_ind_iter+1),:]
+
+                                self.POD_modes[corr][:,ktheta_ind,tfreq_ind,:,:]   = eigmodes
+
+                            else:
+                                ssvd = np.linalg.svd(POD_Mat,full_matrices=False,compute_uv=False)
+
+                            eigval = ssvd**2
+                            self.POD_eigenvalues[corr][ktheta_ind,tfreq_ind,:] = eigval
+
+                    if sort:
+                        sorted_eigind = np.argsort(np.abs(self.POD_eigenvalues[corr]),axis=None)[::-1]
+                        self.sorted_inds[corr]['ktheta'] , self.sorted_inds[corr]['angfreq'], self.sorted_inds[corr]['block'] = np.unravel_index(sorted_eigind,self.POD_eigenvalues['U'][:,:,:].shape)
+
+
+                if len(savefile)>0:
+                    self.variables = {}
+                    self.variables['angfreq'] = angfreq
+                    self.variables['ktheta']   = ktheta
+                    self.variables['r']        = r
+                    self.variables['theta']    = theta
+                    self.variables['y']        = y
+                    self.variables['z']        = z
+                    if not os.path.exists(output_dir):
+                        os.makedirs(output_dir)
+                    savefile = os.path.join(output_dir, savefile)
+                    print("--> Saving to: ",savefile)
+                    objects = [] 
+                    objects.append(self.POD_eigenvalues)
+                    objects.append(self.variables)
+                    if compute_eigen_vectors:
+                        objects.append(self.POD_modes)
+                    if sort:
+                        objects.append(self.sorted_inds)
+                    with open(savefile, 'wb') as f:
+                        for obj in objects:
+                            pickle.dump(obj, f)
```

Comment:
Should we store eigenmodes and eigenvalues separately? What about storing each correlation in a separate file? The eigenmode matrices can be quite large.

Reply:
Added the option to only store n leading eigenmodes to reduce file size. Also changed the ordering of the pickle files so the eigenmodes are read in last, i.e., can be ignored if only eigenvalues are needed.